tls/s2n_signature_algorithms.c

```diff
@@ -50,13 +52,28 @@ static int s2n_signature_scheme_valid_to_offer(struct s2n_connection *conn, cons
 static int s2n_signature_scheme_valid_to_accept(struct s2n_connection *conn, const struct s2n_signature_scheme *scheme)
 {
     POSIX_ENSURE_REF(scheme);
+    POSIX_ENSURE_REF(conn);
 
     POSIX_GUARD(s2n_signature_scheme_valid_to_offer(conn, scheme));
 
     if (scheme->maximum_protocol_version != S2N_UNKNOWN_PROTOCOL_VERSION) {
         POSIX_ENSURE_LTE(conn->actual_protocol_version, scheme->maximum_protocol_version);
     }
 
+    if (conn->actual_protocol_version >= S2N_TLS13
+            || conn->actual_protocol_version == S2N_UNKNOWN_PROTOCOL_VERSION) {
```

Comment:
This is verbatim from the unit tests.

Reply:
How many tests break without it? We really shouldn't be considering S2N_UNKNOWN_PROTOCOL_VERSION here, since outside of the unit tests the protocol versions should be set when this is called.

Reply:
Sorry -- that comment was confusing. I mean that the logic used here is directly from https://github.com/aws/s2n-tls/blob/main/tests/unit/s2n_security_policies_test.c#L831-L838:

```
                /* If scheme will be used for tls1.3 */
                if (max_version == S2N_UNKNOWN_PROTOCOL_VERSION || max_version >= S2N_TLS13) {
                    EXPECT_NOT_EQUAL(scheme->hash_alg, S2N_HASH_SHA1);
                    EXPECT_NOT_EQUAL(scheme->sig_alg, S2N_SIGNATURE_RSA);
                    if (scheme->sig_alg == S2N_SIGNATURE_ECDSA) {
                        EXPECT_NOT_NULL(scheme->signature_curve);
                    }
                }
```

The comparison here is in the event that this function is called before `actual_protocol_version` is known. If the actual_protocol_version is unknown and somehow we are picking a security policy then it had better be one that will work for TLS1.3. --> On the other hand now that I write it out like that it sounds more like this function should never be called when the actual protocol version is unknown. (and is seems you've come to the same while I was writing this :) conclusion).

Okay so I will remove that and make the simplifications you list commented on below -- and modify a bunch of unit tests to set the actual protocol version.

Reply:
Yes, there's a difference between the max version allowed for a signature scheme and the current actual_protocol_version. If the max version for a scheme isn't set, we assume there is no max, so the scheme needs to be valid for TLS1.3. If there is no actual_protocol_version, any protocol version could be chosen including early versions that require SHA1, which means we can't rule out all SHA1 schemes.